lib/faraday/options.rb

```diff
@@ -48,17 +45,32 @@ def clear
     end
 
     # Public
-    def merge(value)
-      dup.update(value)
+    def merge!(other)
+      other.each do |key, other_value|
+        self_value = self.send(key)
+        sub_options = self.class.options_for(key)
+        if sub_options
+          self_value = sub_options.from(self_value)
+          other_value = sub_options.from(other_value)
+          new_value = self_value.merge(other_value)
+        else
+          new_value = other_value
+        end
+        self.send("#{key}=", new_value) unless new_value.nil?
+      end
+      self
+    end
+
+    # Public
+    def merge(other)
+      dup.merge!(other)
     end
-    
+
     # Public
-    def dup
+    def deep_dup
```

Comment:
The deep dup is no longer used after some of the refactoring I did, but I kept it in since it could be useful in the future. I can remove if having "dead" code is not desired.

As to why we can just use dup, it's because `merge!` (bang) recursively calls `merge` (non-bang). Hence, every level will be shallow `dup`ed, which is what a deep_dup would do anyway.